File: baekjoon/binarySearch/2805.py
# 10m

import sys

input = sys.stdin.readline

N, M = map(int,input().split())

treeList = list(map(int,input().split()))

start = 1
end = max(treeList)
maxCnt = 0
maxValue = 0
while start <= end:

    mid = (start + end) // 2
    cutList = []
    for i in treeList:
        if i > mid:
            cutList.append(i - mid)
        else:
            cutList.append(0)

    cutLength = sum(cutList)
    # print(cutLength, M, maxValue,mid)
    if cutLength >= M:
        start = mid + 1
        if maxValue < mid:
            maxValue = mid
    else:
        end = mid - 1

print(maxValue)

File: baekjoon_js/binarySearch/2805.js
("use strict");

const fs = require("fs");
const stdin = (process.platform === "linux"
  ? fs.readFileSync("/dev/stdin").toString()
  : `4 7
20 15 10 17
`
).split("\n");

const input = (() => {
  let line = 0;
  return () => stdin[line++];
})();

const [N, M] = input().split(" ").map(Number);
let woodList = input().split(" ").map(Number);

// Math.max.apply(null, woodList);
let left = 0;
let right = Math.max(...woodList);
let maxValue = 0;

while (left <= right) {
  let mid = parseInt((left + right) / 2);
  let cuttedwood = 0;
  //   woodList.forEach((e) => {
  //     if (e - mid > 0) {
  //       cuttedwood += e - mid;
  //     } else cuttedwood += 0;
  //   });
  for (wood of woodList) {
    if (wood - mid > 0) cuttedwood += wood - mid;
  }

  if (cuttedwood >= M) {
    left = mid + 1;
    if (maxValue < mid) maxValue = mid;
  } else {
    right = mid - 1;
  }
}

console.log(maxValue);
